File: runtime/src/remedies.ts
import { REMEDY_TEMPLATES } from "./remedies.generated.js";

/** Every remedy a tool prints, keyed by its id in `diagnostics/remedies.json`. */
export type RemedyId = keyof typeof REMEDY_TEMPLATES;

/** The placeholders one remedy's fix template names. */
export type RemedyParams<Id extends RemedyId> = Readonly<
  Record<(typeof REMEDY_TEMPLATES)[Id]["params"][number], string | number>
>;

/** The family a remedy belongs to: verifier, trainer-process, runtime or cli. */
export function remedyFamily(id: RemedyId): string {
  return template(id).family;
}

/**
 * The fix text a tool prints for a failure: the template from
 * `diagnostics/remedies.json` with every `{name}` replaced by `params[name]`.
 * An unknown id, a missing parameter or one the template does not name is a
 * programming error and throws.
 */
export function remedy<Id extends RemedyId>(
  id: Id,
  ...args: (typeof REMEDY_TEMPLATES)[Id]["params"] extends readonly []
    ? []
    : [params: RemedyParams<Id>]
): string {
  const entry = template(id);
  const params: Readonly<Record<string, string | number>> = args[0] ?? {};
  const names: readonly string[] = entry.params;
  for (const key of Object.keys(params)) {
    if (!names.includes(key)) {
      throw new TypeError(`remedy ${id} has no parameter ${key}`);
    }
  }
  return entry.fix.replace(/\{([a-z][a-zA-Z0-9]*)\}/gu, (_match, name) => {
    const value = params[name as string];
    if (value === undefined) {
      throw new TypeError(`remedy ${id} needs parameter ${String(name)}`);
    }
    return String(value);
  });
}

function template(id: string): {
  readonly family: string;
  readonly fix: string;
  readonly params: readonly string[];
} {
  if (!Object.hasOwn(REMEDY_TEMPLATES, id)) {
    throw new TypeError(`unknown remedy ${JSON.stringify(id)}`);
  }
  return REMEDY_TEMPLATES[id as RemedyId];
}

File: trainer/src/semantscript_trainer/remedies.py
"""The fix text the trainer prints, from ``diagnostics/remedies.json``.

``scripts/generate-remedies.mjs`` writes the templates into
``remedies_generated.py``; the same source feeds the runtime's and the CLI's
messages and the fix column of ``docs/diagnostics.md``, so they cannot drift.
"""

from __future__ import annotations

import json
import re
from collections.abc import Mapping
from typing import Any, cast

from semantscript_trainer.remedies_generated import REMEDY_TEMPLATES_JSON

REMEDY_TEMPLATES: Mapping[str, Mapping[str, Any]] = cast(
    dict[str, dict[str, Any]], json.loads(REMEDY_TEMPLATES_JSON)
)
_PLACEHOLDER = re.compile(r"\{([a-z][a-zA-Z0-9]*)\}")


def remedy(remedy_id: str, /, **params: object) -> str:
    """Fill one remedy's template; an unknown id or a missing or extra parameter raises."""

    entry = REMEDY_TEMPLATES.get(remedy_id)
    if entry is None:
        raise KeyError(f"unknown remedy {remedy_id!r}")
    names = cast(list[str], entry["params"])
    extra = sorted(set(params) - set(names))
    if extra:
        raise TypeError(f"remedy {remedy_id} has no parameter {', '.join(extra)}")
    missing = [name for name in names if name not in params]
    if missing:
        raise TypeError(f"remedy {remedy_id} needs parameter {', '.join(missing)}")
    return _PLACEHOLDER.sub(lambda match: str(params[match.group(1)]), cast(str, entry["fix"]))


def remedy_family(remedy_id: str, /) -> str:
    """The family one remedy belongs to: verifier, trainer-process, runtime, cli, doctor or editor."""

    entry = REMEDY_TEMPLATES.get(remedy_id)
    if entry is None:
        raise KeyError(f"unknown remedy {remedy_id!r}")
    return cast(str, entry["family"])
